### ISSUE TITLE
Regularely disable append-only mode automatically (optionally)

### ISSUE THREAD

**Issue by rugk** [Ravinou/borgwarehouse#222]
Append-only mode has been introduced in https://github.com/Ravinou/borgwarehouse/issues/160; doc: https://borgwarehouse.com/docs/user-manual/repositories/#append-only-mode

However, the issue obviously is this can fill up your disk space.

Can you somehow add an option to enable or disable it automatically, say… every 30days? (And re-enables it after pruning or after some interval again like maybe 10h or whatever – where the backup client is scheduled to do it's backup.) So you have at least 30 days of backups?

Threat model assumes a compromised client, but one, where you would at least in some interval notice this. Like a ransomware will hopefully after… say… 30 days at most show you a notice to pay.
Of course, there is a risk that "real pruning" happens at some time, when the data has been deleted, but AFAIK, there is not a lot you can do here. The risk is always there… because… even if you don't see it ransomware could have encrypted/deleted everything already.

## Alternatives

The alternative is kinda the same thing: You might manually check this whether backups have been made/the system is okay and then disable append-only,
So depending on your threat model and how sophisticated you guess your attacker is, the "automatic append-only disable" mode may be sufficient IMHO and be better than not having append-only mode at all. Because realistically, this is the more relativistic alternative…

## More ideas
Thinking about it, you could have a simple "sanity check" (or some would make market it and solve it with AI haha…) that just checks before the mode is disabled, whether:
* much da has been pruned recently
* any other criteria that could be a tell for a compromise of the client?

In such a case (even when this option is not enabled), there could be some warning (/triangle) or so being displayed like: "In the last 7 [30|60] days, more than 6GB of data has been deleted. This could be generally a nice feature.
AFAIK this is how [Nextcloud's ransomware protection app](https://apps.nextcloud.com/apps/ransomware_protection) [kinda works](https://nextcloud.com/blog/nextcloud-presents-ransomware-protection-app/).

**Comment by Ravinou** [comment]
Thanks for your comment. As I was developing the feature, I was thinking of the same thing as you, but in reverse 😄

**In other words, a temporary deactivation of append-only mode. That way you can do your pruning and not forget to reactivate the option afterwards, for example.**

But, an option to deactivate the append-only mode over a time window might be worth considering. Still, I can't see the point of such a feature. Automatically disabling append-only mode would mean that your client would have to do the prune in the same time window, and automatically too. It all sounds very flawed to me 🙄

**Comment by rugk** [comment]
Yeah, I meant exactly such option.

**Comment by Forceu** [comment]
How about having a button in the interface instead for compacting? This can be done server-side without a passphrase.

**Comment by rugk** [comment]
Hmm but borg compact is not prune, but does this work in a fully automated fashion maybe (so not client-server coordination needs to be done)?

At regular intervals invoke the following "algorithm":
1. Disable append-only mode. 
2. Invoke compact.
3. Enable append-only mode afterwards.

Clients can then run prune as usual and the server just "once upon a month" or so, cleans up what needs to be cleaned up while keeping the repo in append-only mode otherwise?

**Comment by hozza** [comment]
Borg Base handles this by allowing the client to to specify prune but in append only mode it will not actually delete those repos marked as "to be pruned". Then periodically (e.g. every 6 weeks) the server side runs compact which actually deletes the specified prunes. Essentially converting the append only option into a safe "delayed deletion" determined by the server side.

**Comment by rugk** [comment]
I agree! IIRC when I wrote my comments before borg did not had the `compact` command or I did not have it in mind yet…

**Comment by Forceu** [comment]
Any update on this? Currently I am manually disabling append-only, then running `borg compact [repo]` via ssh and then enabling append-only again.

Borgbase for example has a button for this in the UI:

![Image](https://github.com/user-attachments/assets/9b47b3d4-ccbb-4e8a-b638-88de6e81ba27)


It would be great to have something like this as well in the UI. If I have the time, I might look into creating a PR, however I don't have a lot experience with TS.

**Comment by Forceu** [comment]
I am currently working on a PR to add the feature #468

**Comment by hozza** [comment]
> I am currently working on a PR to add the feature [#468](https://github.com/Ravinou/borgwarehouse/pull/468)

@Forceu I just saw in a Borgbase update email that version 2.0.0b17 (2025-05-23) will [change how this permission works](https://borgbackup.readthedocs.io/en/2.0.0b19/changes.html#version-2-0-0b17-2025-05-23). Not sure if this will effect your PR?

i.e. `BORG_REPO_PERMISSIONS=all|no-delete|write-only|read-only, #8823`

**Comment by sdinot** [comment]
A more subtle strategy is possible: declare two SSH keys.
1. One key for which “append-only” access is configured. This key is used by the backup script that runs automatically.
2. A key for which normal access is configured. This key is used by the administrator to launch manual purge commands from time to time (only the administrator knows the password for this key; the script cannot use it, even in the event of a configuration error).

This strategy is documented here:
https://borgbackup.readthedocs.io/en/stable/usage/notes.html#append-only-mode-forbid-compaction
> In addition, borg serve can act as if a repository is in append-only mode with its option `--append-only`. This can be very useful for fine-tuning access control in `.ssh/authorized_keys`:
>
> ```conf
> command="borg serve --append-only ..." ssh-rsa <key used for not-always-trustworthy backup clients>
> command="borg serve ..." ssh-rsa <key used for backup management>
> ```

I use this approach on my backup server.

Borgbase offers the same feature, which can be configured via the WebUI:

<img width="700" height="471" alt="Image" src="https://github.com/user-attachments/assets/fca31948-f1e2-4ac9-a5b8-ef40ad46fc56" />
